app/controllers/controls/views/gradients.js

```diff
@@ -0,0 +1,110 @@
+var log = require('log');
+
+var normalizedCenterX;
+var normalizedCenterY;
+var colors;
+var Gradient;
+var gradientTypes;
+var selectedGradient;
+
+/**
+ * The scoped constructor of the controller.
+ **/
+(function constructor() {
+  normalizedCenterX = 0.5;
+  normalizedCenterY = 0.5;
+  colors = [ 'red', 'blue' ];
+  Gradient = { RADIAL: 0, LINEAR: 1 };
+  gradientTypes = [ 'radial', 'linear' ];
+  selectedGradient = Gradient.RADIAL;
+})();
+
+function updateGradient() {
+	var size = $.gradientView.rect;
+	var minDimension = (size.width < size.height) ? size.width : size.height;
+	var centerPoint = {
+		x: size.width * normalizedCenterX,
+		y: size.height * normalizedCenterY,
+	};
+	
+  var startRadius = (minDimension / 2) * ($.startRadiusSlider.value / 100);
+  var endRadius = (minDimension / 2) * ($.endRadiusSlider.value / 100);
+  
+  var gradient = {
+		type: gradientTypes[selectedGradient]
+	};
+  
+  // Linear gradients support colors with offsets and start-point / end-point
+  // Radial gradients support raw colors, start-radius / end-radius and backfill-start / backfill-end
+  // Read more: http://docs.appcelerator.com/platform/latest/#!/api/Titanium.UI.View-property-backgroundGradient
+  if (selectedGradient == Gradient.LINEAR) {
+    var startPoint = { x: precisionRound($.startRadiusSlider.value , -1) + '%', y: '50%' };
+    var endPoint = { x: precisionRound($.endRadiusSlider.value, -1) + '%', y: '50%' };
+
+    gradient.colors = [{ color: colors[0], offset: 0.0 }, { color: colors[1], offset: 1.0 }];
+    gradient.startPoint = startPoint;
+    gradient.endPoint = endPoint;
+
+    log.args('Linear gradient updated: ' + JSON.stringify(centerPoint) + ', start-point: ' + JSON.stringify(startPoint) + ', end-point: ' + JSON.stringify(endPoint));
+  } else {
+    gradient.startPoint = centerPoint;
+    gradient.endPoint = centerPoint;
+    gradient.startRadius = startRadius;
+    gradient.endRadius = endRadius;
+    gradient.backfillStart = $.startFillSwitch.value;
+    gradient.backfillEnd = $.endFillSwitch.value;
+    gradient.colors = colors;
+
+    log.args('Radial gradient updated: ' + JSON.stringify(centerPoint) + ', start-radius: ' + startRadius + ', end-radius: ' + endRadius);
+  }
+  
+  $.gradientView.backgroundGradient = gradient;  
+}
+
+function handleTouchMove(e) {
+  var size = $.gradientView.rect;
+
+  if (OS_ANDROID && Ti.UI.defaultunit !== 'px') {
+    e.x /= Ti.Platform.displayCaps.logicalDensityFactor;
+    e.y /= Ti.Platform.displayCaps.logicalDensityFactor;
+  }
```

Comment:
For your info, we have a platform parity issue here. Both Android and iOS ignore the units set in the "tiapp.xml" file.

Android always provides touch coordinates in pixels.
iOS always provides touch coordinates in points.
_(I haven't checked the iOS code, but that looks like what's happening.)_

We should probably update the docs.
Resolving this parity issue would be a breaking change though. Something to consider for Titanium 8.0.0.

Reply:
Oh shoot. I messed up. There is no `Ti.UI.defaultunit` API. We need to fetch the default unit like this instead...
```javascript
var defaultUnit = Ti.App.Properties.getString("ti.ui.defaultunit", "dp");
```

Then we can do the following work-around on both Android and iOS...
```javascript
if (Ti.Platform.name === "android") {
	event.x = Ti.UI.convertUnits(event.x + "px", defaultUnit);
	event.y = Ti.UI.convertUnits(event.y + "px", defaultUnit);
} else if ((Ti.Platform.name === "iOS") || (Ti.Platform.name === "iPhone OS")) {
	event.x = Ti.UI.convertUnits(event.x + "dip", defaultUnit);
	event.y = Ti.UI.convertUnits(event.y + "dip", defaultUnit);
}
```

I've confirmed that the above works for default units `dp`, `px`, and `mm`.

Reply:
I wrote this up as a ticket here: [TIMOB-25839](https://jira.appcelerator.org/browse/TIMOB-25839)

Reply:
Thanks Josh! I remember a guide we had on that topic as well, let me find that as well. But for this example, it looks okay :-)